### ISSUE TITLE
dir_id must be an integer in box_fetch but not box_ls() 

### ISSUE THREAD

**Issue by tbuckl** [brendan-r/boxr#80]
hello! this is a great package. 

i noticed that for:
```
box_fetch(dirid)
```
dirid must be an integer 

but not for:
```
box_ls(dirid)
#or
box_setwd()
```

but that this is not in the docs. if i find the time to PR will change. 

otherwise logging here.

**Comment by ijlyttle** [comment]
Thanks - I will look into this more.

I think the right answer for R is that directory- and file-id's should be doubles or character.

Ultimately, the id's are character based because they are transmitted as a part of a web-service call. I don't think that box assigns any meaning to them and I don't think that we are adding or subtracting them.

Box uses numeric characters only, so they have the *appearance* of integers. Here's where I see the problem in R - the size of the "integers" that Box uses is larger than the maximum size for integers in R, so we approximate the integer using a double, which seems to work for now.

Again, let me have a look - I think the right thing to do is to coerce all such inputs to character type.

**Comment by nathancday** [comment]
I'm gonna work on this issue

**Comment by ijlyttle** [comment]
Hi @nathancday, thanks! 

My gut feeling is that we whenever we take in an id, we want to coerce it to `character`, then work with it internally as `character`. Ultimately, it gets sent to box as `character`.  In other words, box uses ids that are `character` strings that happen to contain only digits.

The only way this makes sense as an `integer` is if we are performing arithmetic operations on it, which I don't think we are doing.

**Comment by nathancday** [comment]
I agree 100% and it looks adding `as.character()` would be easy to add in universally.

FWIW it looks like the tripping part here was in the internal function `dirTreeRecusive()` building a data.frame badly. 

From lines 324-330:
```
  dir_row <- 
    data.frame(
      type = "folder", id = dir_id, sequence_id = 0,
      etag = 0, name = "",
      local_dir = local_dir,
      path = "~"
    )
```
Without `stringAsFactors = FALSE` the character were being *coherced* to a factor silently and causing the problem only in `box_fetch()`.

Since `box_ls` and `box_setwd` have been playing nice with characters and doubles do we want to add the coercion for factors? I think it might be good to continue failing on factors, but with a better error message. Your thoughts?

**Comment by nathancday** [comment]
I like the idea of `as_id()` long-term. I just submitted the fix for now and test to check.

Good luck on the testing, have you started working on the Pkgdown yet I see the branch `gh-pages`, so wanted to check

**Comment by ijlyttle** [comment]
Thanks - I'll have a look!

re: pkgdown: yes - you can see a first version of it at https://ijlyttle.github.io/boxr/ - I am working with @brendan-r to get things automated for his repository.

**Comment by nathancday** [comment]
Awesome looking good!

**Comment by nathancday** [comment]
should we merge this fix for short-term, eg 3.5 release, and open a new issue for `as_id()` in `3.6` goals/milestones?

**Comment by ijlyttle** [comment]
I think so - given that we will need to wait for rio, it might happen before we go again to CRAN :)

**Comment by nathancday** [comment]
based on rio's past CRAN release schedule it is due for another soon. The version is at 5.18 and the last CRAN published on was 5.16, so we might not be waiting too long

**Comment by brendan-r** [comment]
Extremely belated thank-you to the two of you for fixing this! (Reading through the issue revived a vague sense of dread I had about this several years ago.) :heart:
